File: hivemind/types/task.py
import hashlib
import json
from enum import Enum
from pydantic import BaseModel


class TaskStatus(Enum):
    PENDING = 0
    RUNNING = 1
    COMPLETED = 2
    FAILED = -1


class Task(BaseModel):
    id: str
    description: str
    dependencies: list[str] = []
    status: TaskStatus = TaskStatus.PENDING
    result: str | None = None
    error: str | None = None  # v1.9: error message when failed
    speculative: bool = False
    role: str | None = None  # Optional agent role: research, code, analysis, critic
    retry_count: int = 0  # v1.7: critic retries

    def to_dict(self) -> dict:
        """Return all fields as JSON-safe dict."""
        return {
            "id": self.id,
            "description": self.description,
            "dependencies": list(self.dependencies),
            "status": self.status.value if hasattr(self.status, "value") else str(self.status),
            "result": self.result,
            "error": self.error,
            "speculative": self.speculative,
            "role": self.role,
            "retry_count": self.retry_count,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Task":
        """Reconstruct Task from dict. Parse status back to TaskStatus enum."""
        status = data.get("status", TaskStatus.PENDING)
        if isinstance(status, int):
            task_status = TaskStatus(status)
        elif isinstance(status, str):
            name_to_status = {
                "PENDING": TaskStatus.PENDING,
                "RUNNING": TaskStatus.RUNNING,
                "COMPLETED": TaskStatus.COMPLETED,
                "FAILED": TaskStatus.FAILED,
                "0": TaskStatus.PENDING,
                "1": TaskStatus.RUNNING,
                "2": TaskStatus.COMPLETED,
                "-1": TaskStatus.FAILED,
            }
            task_status = name_to_status.get(status.upper(), TaskStatus.PENDING)
        else:
            task_status = TaskStatus.PENDING
        return cls(
            id=data["id"],
            description=data.get("description", ""),
            dependencies=list(data.get("dependencies", [])),
            status=task_status,
            result=data.get("result"),
            error=data.get("error"),
            speculative=data.get("speculative", False),
            role=data.get("role"),
            retry_count=data.get("retry_count", 0),
        )

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, raw: str) -> "Task":
        return cls.from_dict(json.loads(raw))

    def checksum(self) -> str:
        """SHA256 of to_json(); used to detect state drift between nodes."""
        return hashlib.sha256(self.to_json().encode("utf-8")).hexdigest()

File: worker/src/types/task.rs
//! Task type compatible with Python Task (to_dict/from_dict).

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum TaskStatus {
    #[default]
    Pending = 0,
    Running = 1,
    Completed = 2,
    Failed = -1,
}

impl TaskStatus {
    pub fn from_i64(v: i64) -> Self {
        match v {
            0 => TaskStatus::Pending,
            1 => TaskStatus::Running,
            2 => TaskStatus::Completed,
            -1 => TaskStatus::Failed,
            _ => TaskStatus::Pending,
        }
    }
}

impl serde::Serialize for TaskStatus {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_i64(*self as i64)
    }
}

impl<'de> serde::Deserialize<'de> for TaskStatus {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let v = i64::deserialize(deserializer)?;
        Ok(Self::from_i64(v))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    pub id: String,
    pub description: String,
    #[serde(default)]
    pub dependencies: Vec<String>,
    #[serde(default)]
    pub status: TaskStatus,
    pub result: Option<String>,
    pub error: Option<String>,
    #[serde(default)]
    pub speculative: bool,
    pub role: Option<String>,
    #[serde(default)]
    pub retry_count: u32,
}

impl Task {
    pub fn status_value(&self) -> i64 {
        self.status as i64
    }
}
